[Rust] Add first stage of updating and rewriting Rust bindings.

jroesch opened the pull request:
I plan on sending an RFC tomorrow on the greater refactor but my goal is to restructure the current Rust code into 4 crates:
- `tvm-sys` containing the lowest level bindings to `libruntimetvm.dylib`
- `tvm-rt` a layer which provides a ergonomic Rust layer over the primitive concepts exposed by `tvms-sys`.
- `tvm` a high level crate for using the full TVM library including the compiler components. 
- `tvm-graph-rt` previously the `runtime` a library a ABI compatible Rust implementation of the graph runtime. 

I was attempting to refactor this in one go, but have found my current branch is rapidly reaching a few thousand lines.

I will first send it in pieces to be reviewed and merged, and then follow up with connecting PRs to put the pieces together, turn on CI, and replace the existing code. 
@binarybana @robo-corg @mwillsey @vegaluisjose @MarisaKirisame @mbrookhart @tqchen

tqchen commented:
cc @nhynes

nhynes commented on rust/tvm-sys/src/array.rs:
```diff
@@ -0,0 +1,62 @@
+                        bits: 8 * mem::size_of::<$type>() as u8,
+                        lanes: 1,
+                    },
+                    shape: arr.shape().as_ptr() as *const i64 as *mut i64,
+                    strides: arr.strides().as_ptr() as *const isize as *mut i64,
```
looks like you can use [as_mut_ptr](https://docs.rs/ndarray/0.13.1/ndarray/struct.ArrayBase.html#method.as_mut_ptr) now

nhynes commented on rust/tvm-sys/src/byte_array.rs:
```diff
@@ -0,0 +1,64 @@
+use std::os::raw::c_char;
+
+use crate::ffi::TVMByteArray;
```
this is stylistic, but if you truly mean to never newtype this struct, you can
```rust
pub use crate::ffi::TVMByteArray as ByteArray;
```

nhynes commented on rust/tvm-sys/src/byte_array.rs:
```diff
@@ -0,0 +1,64 @@
+
+use crate::ffi::TVMByteArray;
+
+/// A struct holding TVM byte-array.
```
```suggestion
/// A struct holding TVM byte-array.
```
is it really holding? looks pretty direct to me. what interface are you going for? maybe do a newtype wrapper? dunno

jroesch replied:
This comment also comes from the original code.

nhynes commented on rust/tvm-sys/src/context.rs:
````diff
@@ -0,0 +1,293 @@
+///```
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct TVMDeviceType(pub i64);
````
I would probably opt or this not to expose the internal repr. Why not make it a `repr(i64)` enum, actually? Then you can just `as` it around as necessary. If you don't mind using rust nightly, the [proper](https://docs.rs/proper/0.1.5/proper/) crate might be of use.

nhynes commented on rust/tvm-sys/src/context.rs:
```diff
@@ -0,0 +1,293 @@
+impl From<ffi::DLDeviceType> for TVMDeviceType {
+    fn from(device_type: ffi::DLDeviceType) -> Self {
+        match device_type {
+            ffi::DLDeviceType_kDLCPU => TVMDeviceType(1),
```
yeah I'd definitely consider making this a `repr(i64)` enum. I get that it's easy to catch bugs here, but I fear that they will eventually occur.

jroesch replied:
Yeah sounds good, another example of leaving existing code alone and rewriting everything in one go :)

nhynes commented on rust/tvm-sys/src/context.rs:
```diff
@@ -0,0 +1,293 @@
+            "metal" => TVMDeviceType(8),
+            "vpi" => TVMDeviceType(9),
+            "rocm" => TVMDeviceType(10),
+            _ => panic!("{:?} not supported!", type_str),
```
nothing to be done about `"rpc"` right?

jroesch replied:
I don't think RPC exists in the device types anymore if you look at the dlpack.h header.

nhynes commented on rust/tvm-sys/src/context.rs:
```diff
@@ -0,0 +1,293 @@
+use std::convert::TryFrom;
+use std::str::FromStr;
+use thiserror::Error;
+
```
why this line break?

nhynes commented on rust/tvm-sys/src/context.rs:
```diff
@@ -0,0 +1,293 @@
+
+use std::convert::TryFrom;
+use std::str::FromStr;
+use thiserror::Error;
```
why is this external crate not grouped with the other external crate?

nhynes commented on rust/tvm-sys/src/datatype.rs:
```diff
@@ -0,0 +1,179 @@
+ * under the License.
+ */
+
+use std::any::TypeId;
+
+use std::convert::TryFrom;
+use std::str::FromStr;
+
+use crate::packed_func::RetValue;
+
+use thiserror::Error;
+
+use crate::ffi::DLDataType;
```
I'm having a really hard time picking up on your import sorting style :)

jroesch replied:
I will turn on the `rustfmt` flag for them. I'll do a cleanup pass have been very quickly hacking and trying to cut the PR up before its too large :).

nhynes commented on rust/tvm-sys/src/datatype.rs:
```diff
@@ -0,0 +1,179 @@
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct DataType {
+    pub code: u8,
+    pub bits: u8,
+    pub lanes: u16,
```
I do wonder if these have to be `pub` when you're providing nice getters and `new`

jroesch replied:
Good point, I will try and change this.

nhynes commented on rust/tvm-sys/src/datatype.rs:
```diff
@@ -0,0 +1,179 @@
+        }
+
+        let mut type_lanes = type_str.split('x');
+        let typ = type_lanes.next().expect("Missing dtype");
```
```suggestion
        let typ = type_lanes.next().ok_or(ParseTvmError::MissingDataType)?;
```

jroesch replied:
👍

nhynes commented on rust/tvm-sys/src/datatype.rs:
```diff
@@ -0,0 +1,179 @@
+            return write!(f, "bool");
+        }
+        let mut type_str = match self.code {
+            0 => "int",
```
these magic numbers are used in two places (the other is `is_type`). I'd probably declare constants

jroesch replied:
Sounds good.

nhynes commented on rust/tvm-sys/src/lib.rs:
```diff
@@ -0,0 +1,54 @@
+pub mod value;
+
+pub use byte_array::ByteArray;
+pub use context::{Context, TVMDeviceType};
```
why not `DeviceType` if you already stripped the prefix from `ByteArray`?

jroesch replied:
I will do that as well, I think we can just rely on Rust namespacing in general :)

nhynes commented on rust/tvm-sys/src/packed_func.rs:
```diff
@@ -0,0 +1,380 @@
+        $fn(&[$($args.into(),)+])
+    };
+    ($fn:expr) => {
+        $fn(&Vec::new())
```
```suggestion
        $fn(&[])
```
neither allocates, of course, but this feels more direct

nhynes commented on rust/tvm-sys/src/packed_func.rs:
```diff
@@ -0,0 +1,380 @@
+        Str(&'static CStr),
+    },
+    match value {
+        TVMTypeCode_kTVMBytes => { Bytes(*(value.v_handle as *const TVMByteArray)) }
```
can you make a note somewhere that the C++ lib retains ownership of all objects? Otherwise, I'd expect a `Drop` impl.

jroesch replied:
Yeah I think this code is from the previous iteration, there are a couple ownership issues we probably need to clean up. I've been trying to radically simplify the interface.

nhynes commented on rust/tvm-sys/src/value.rs:
```diff
@@ -0,0 +1,95 @@
+ * under the License.
+ */
+
+use std::str::FromStr;
+
+use thiserror::Error;
+
+use crate::ffi::*;
```
please sort imports? I personally like ordering them in block of specificity, so

```rust
#[macro_rules]
extern crate macro_provider;

mod my_mod;

use std::*;

use extern_crate::*;

use crate::*;
```
And then I turn on `merge-imports` in the `.rustfmt.toml` to keep things tidy.

jroesch replied:
I've started to rely on just using `rustfmt` instead of paying attention to my coding style. I will flip the flag and organize them.

jroesch replied:
I applied the splits like so, everywhere I saw it was different.

jroesch commented:
@nhynes thanks for the review. I opened this at like 2:30am last night before going to bed. Will provide some more context.

The goal of my refactor to is to turn on the new C-ABI compatible object system in TVM. I started by trying to simplify the split between interfaces and better match TVM's linking model. I found some of the code using the low-level `ffi::{X, Y}` types and some using higher-level types, this first patch just combines the overlap without fully changing things to my liking as in not to break everything. 

What I first did is combine some overlapping code that I found in `tvm-frontend` and `tvm-common` and then have been working on rectifying the top-level layers. If you want to see end-to-end state you can see my WIP version at https://github.com/jroesch/tvm/tree/rust-fe. 

A cool not yet polished thing is the ability to write an out of tree Rust pass. See https://github.com/jroesch/tvm/tree/rust-fe/rust/out-of-tree. 

I would like to land the initial layers in tree so I can start to turn on CI and incrementally land instead of writing a single 10kloc PR. Hoping to do multiple passes on the naming, interface, etc.

ehsanmok commented:
These changes make very good sense. I'm glad Rust is receiving more love and attention.

jroesch commented:
Yeah @ehsanmok I'm hoping to reduce the effort for us to use/consume. My goal is effectively to get to:

```
#[external_pass]
fn pass(func: relay::Function, ...) -> relay::Function { ... }
```

At the high level and reduce some duplication and mirror the TVM bindings.

jroesch commented:
I addressed all the feedback except a few things that don't work like changing to `get_mut_ptr()` which doesn't work due to the array being borrowed and not mutably borrowed. Looking into that DLCode I think having a from instance is not very safe given the lifetime of the DLTensor is not tied to the lifetime of the NDArray in anyway.

jroesch commented:
@nhynes @tqchen @ehsanmok can someone approve the changes so we can land this? it should be good to go, if you have future follow up work I'm happy to do on another branch.

adelbertc commented on rust/tvm-sys/build.rs:
```diff
@@ -0,0 +1,117 @@
+// use cmake::Config;
+
+// fn main() {
+//     if !Path::new("tvm/.git").exists() {
```
Delete code?

jroesch replied:
I will remove from this PR, I need to restore once I land the second PR.

adelbertc commented on rust/tvm-sys/build.rs:
```diff
@@ -0,0 +1,117 @@
+        println!("cargo:rerun-if-env-changed=TVM_HOME");
+        // println!("cargo:rustc-link-lib=dylib=tvm_runtime");
+        // TODO: move to core
+        // println!("cargo:rustc-link-lib=dylib=tvm_runtime");
```
Delete?

jroesch replied:
I will restore in follow up PR.

adelbertc commented on rust/tvm-sys/src/context.rs:
```diff
@@ -0,0 +1,285 @@
+use crate::packed_func::{ArgValue, RetValue};
+
+use thiserror::Error;
+use anyhow::Result;
```
My take is that libraries should not be using `anyhow::Error` and instead should use the stdlib mechanisms so as to not be prescriptive. For example the pattern used in [reqwest](https://docs.rs/reqwest/0.10.4/src/reqwest/error.rs.html#12-14) seems common in my experience. Just my 2c.

jroesch replied:
I think we should revisit this when I land the higher-level errors, before these were all using failure and and I tried to just 1-1 map failure use cases to thiserror/anyhow. One of the challenges is handling dynamic errors produced by TVM.

adelbertc commented on rust/tvm-sys/src/errors.rs:
```diff
@@ -0,0 +1,46 @@
+ * under the License.
+ */
+
+use thiserror::Error;
```
See my previous comment on using `thiserror`.

robo-corg commented on rust/tvm-sys/src/datatype.rs:
```diff
@@ -0,0 +1,188 @@
+use thiserror::Error;
+
+
+const DL_INT_CODE: u8 = 0;
+const DL_UINT_CODE: u8 = 1;
+const DL_FLOAT_CODE: u8 = 2;
+const DL_HANDLE: u8 = 3;
```
Should these be part of the public api?

jroesch replied:
No which is why they are private.

robo-corg replied:
Should DataType::code be public then?

jroesch commented on rust/tvm-sys/src/context.rs:
````diff
@@ -0,0 +1,285 @@
+/// println!("device is: {}", cpu);
+///```
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, N)]
+#[repr(i64)]
+pub enum DeviceType {
+    CPU = 1,
+    GPU,
+    CPUPinned,
+    OpenCL,
+    Vulkan,
+    Metal,
+    VPI,
+    ROCM,
+    ExtDev,
+}
````
`EnumN` allows me to convert back from the raw integers produced by the raw FFI with validation.

jroesch replied:
The `tvm-sys` stands alone from all other crates right now. The goal is to fix the above issues in the higher level crates by wrapping standard low-level constructs with better designs and then incrementally bubbling up to other crates, see my other open PR for more examples.

robo-corg replied:
Will do! The changes I suggested I think still allow this conversion without the extra crate unless I am missing something.

robo-corg replied:
Spoiler: I was missing something and you were right!

binarybana commented on rust/tvm-sys/Cargo.toml:
```diff
@@ -0,0 +1,35 @@
+
+[dependencies]
+thiserror = "^1.0"
+anyhow = "^1.0"
```
I'm not sure we should bring in `anyhow` into a library crate (AFAIK it's form for libraries to pollute their API surface by exposing opinionated error types). But given that we only use it in one place, we can remove it later.

binarybana commented on rust/tvm-sys/src/packed_func.rs:
```diff
@@ -0,0 +1,380 @@
+        }
+    }
+}
+
```
Putting a proper rustdoc item on this macro would be appreciated given the complexity and the relationship to the one above and below. Even if these are internal, to help people hack on this code in the future.

binarybana commented on rust/tvm-sys/src/packed_func.rs:
```diff
@@ -0,0 +1,380 @@
+    }
+}
+
+TVMPODValue! {
```
Missing rustdoc note here too.

jroesch commented:
@binarybana I am planning rapid fire follow up PRs I will fold the feedback into the next one. @nhynes if you have any follow up feedback I can send other PRs, @tqchen and I will merge so I can continue working on the follow up.
